CI: control via commit message (#5074)
ComputationalRadiationPhysics/picongpu: docs/source/dev/ci.rst

.. _development-ci:

Continuous Integration
======================

.. sectionauthor:: René Widera

What is tested?
---------------

The CI is compiling tests with different compilers and boost versions.
We compile examples, tests, and benchmark input sets from ``share/picongpu``.
Unit tests for PMacc from ``include/pmacc/test`` and examples under ``share/pmacc`` will be compiled and executed on the corresponding compute device.

> PIConGPU unit test from ``share/picongpu/unit`` will be compiled and executed  on the corresponding compute device.

Reviewer: ```suggestion
PIConGPU unit test from ``share/picongpu/unit`` will be compiled and executed on the corresponding compute device.
```

PICMI from ``share/picongpu/pypicongpu`` will be compiled too.

Compiler chains
---------------

The CI is testing different version of

* ``nvcc`` with ``g++`` for NVIDIA CUDA compute devices
* ``clang`` with HIP for AMD ROCm compute devices
* ``clang`` as NVIDIA CUDA compiler for NVIDIA CUDA devices
* ``clang`` and ``g++`` for the serial alpaka accelerator to target x86-CPU compute devices

Test PR
-------

Pull requests will perform all tests described above.
By default PIConGPU input sets will only compile a reduced set of all setups.
Only the first input variation from ``cmakeFlags`` will be compiled

> There is no guarantee that each setup is compiled with at least on version of the compiles chains above.

Reviewer: ```suggestion
There is no guarantee that each setup is compiled with at least one version of the compiler chains above.
```

> It is guaranteed that each compiler version of all compiler chains is compiling at least on PIConGPU input set.

Reviewer: ```suggestion
It is guaranteed that each compiler version of all compiler chains is compiling at least one PIConGPU input set.
```
